File: gitlfs-server/src/main/java/ru/bozaro/gitlfs/server/LockManager.java
package ru.bozaro.gitlfs.server;

import ru.bozaro.gitlfs.common.LockConflictException;
import ru.bozaro.gitlfs.common.VerifyLocksResult;
import ru.bozaro.gitlfs.common.data.Lock;
import ru.bozaro.gitlfs.common.data.Ref;

import javax.annotation.CheckForNull;
import javax.annotation.Nonnull;
import javax.servlet.http.HttpServletRequest;
import java.io.IOException;
import java.util.List;

public interface LockManager {

  @Nonnull
  LockRead checkDownloadAccess(@Nonnull HttpServletRequest request) throws IOException, ForbiddenError, UnauthorizedError;

  @Nonnull
  LockWrite checkUploadAccess(@Nonnull HttpServletRequest request) throws IOException, ForbiddenError, UnauthorizedError;

  interface LockRead {
    @Nonnull
    List<Lock> getLocks(@CheckForNull String path, @CheckForNull String lockId, @CheckForNull Ref ref) throws IOException;
  }

  interface LockWrite extends LockRead {
    @Nonnull
    Lock lock(@Nonnull String path, @CheckForNull Ref ref) throws LockConflictException, IOException;

    @CheckForNull
    Lock unlock(@Nonnull String lockId, boolean force, @CheckForNull Ref ref) throws LockConflictException, IOException;

    @Nonnull
    VerifyLocksResult verifyLocks(@CheckForNull Ref ref) throws IOException;
  }
}

File: gitlfs-server/src/main/kotlin/ru/bozaro/gitlfs/server/LockManager.kt
package ru.bozaro.gitlfs.server

import jakarta.servlet.http.HttpServletRequest
import ru.bozaro.gitlfs.common.LockConflictException
import ru.bozaro.gitlfs.common.VerifyLocksResult
import ru.bozaro.gitlfs.common.data.Lock
import ru.bozaro.gitlfs.common.data.Ref
import java.io.IOException

interface LockManager {
    @Throws(IOException::class, ForbiddenError::class, UnauthorizedError::class)
    fun checkDownloadAccess(request: HttpServletRequest): LockRead

    @Throws(IOException::class, ForbiddenError::class, UnauthorizedError::class)
    fun checkUploadAccess(request: HttpServletRequest): LockWrite
    interface LockRead {
        @Throws(IOException::class)
        fun getLocks(path: String?, lockId: String?, ref: Ref?): List<Lock>
    }

    interface LockWrite : LockRead {
        @Throws(LockConflictException::class, IOException::class)
        fun lock(path: String, ref: Ref?): Lock

        @Throws(LockConflictException::class, IOException::class)
        fun unlock(lockId: String, force: Boolean, ref: Ref?): Lock?

        @Throws(IOException::class)
        fun verifyLocks(ref: Ref?): VerifyLocksResult
    }
}
